File: ptm_task/src/infrastructure/store/mapper/note_mapper.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package mapper

import (
	"encoding/json"

	"github.com/serp/ptm-task/src/core/domain/entity"
	"github.com/serp/ptm-task/src/infrastructure/store/model"
	"gorm.io/datatypes"
)

type NoteMapper struct{}

func NewNoteMapper() *NoteMapper {
	return &NoteMapper{}
}

func (m *NoteMapper) ToEntity(noteModel *model.NoteModel) *entity.NoteEntity {
	if noteModel == nil {
		return nil
	}

	note := &entity.NoteEntity{
		BaseEntity: entity.BaseEntity{
			ID:        noteModel.ID,
			CreatedAt: noteModel.CreatedAt.UnixMilli(),
			UpdatedAt: noteModel.UpdatedAt.UnixMilli(),
		},
		UserID:   noteModel.UserID,
		TenantID: noteModel.TenantID,

		TaskID:    noteModel.TaskID,
		ProjectID: noteModel.ProjectID,

		Content:      noteModel.Content,
		ContentPlain: noteModel.ContentPlain,

		IsPinned:     noteModel.IsPinned,
		ActiveStatus: noteModel.ActiveStatus,
	}

	note.Attachments = m.jsonToAttachments(noteModel.Attachments)

	return note
}

func (m *NoteMapper) ToModel(note *entity.NoteEntity) *model.NoteModel {
	if note == nil {
		return nil
	}

	noteModel := &model.NoteModel{
		BaseModel: model.BaseModel{
			ID: note.ID,
		},
		UserID:   note.UserID,
		TenantID: note.TenantID,

		TaskID:    note.TaskID,
		ProjectID: note.ProjectID,

		Content:      note.Content,
		ContentPlain: note.ContentPlain,

		IsPinned:     note.IsPinned,
		ActiveStatus: note.ActiveStatus,
	}

	noteModel.Attachments = m.attachmentsToJSON(note.Attachments)

	return noteModel
}

func (m *NoteMapper) ToEntities(models []*model.NoteModel) []*entity.NoteEntity {
	if models == nil {
		return nil
	}

	entities := make([]*entity.NoteEntity, 0, len(models))
	for _, model := range models {
		if entity := m.ToEntity(model); entity != nil {
			entities = append(entities, entity)
		}
	}

	return entities
}

func (m *NoteMapper) ToModels(entities []*entity.NoteEntity) []*model.NoteModel {
	if entities == nil {
		return nil
	}

	models := make([]*model.NoteModel, 0, len(entities))
	for _, entity := range entities {
		if model := m.ToModel(entity); model != nil {
			models = append(models, model)
		}
	}

	return models
}

func (m *NoteMapper) attachmentsToJSON(attachments []entity.NoteAttachment) datatypes.JSON {
	if attachments == nil {
		return datatypes.JSON([]byte("[]"))
	}

	bytes, err := json.Marshal(attachments)
	if err != nil {
		return datatypes.JSON([]byte("[]"))
	}

	return datatypes.JSON(bytes)
}

func (m *NoteMapper) jsonToAttachments(jsonData datatypes.JSON) []entity.NoteAttachment {
	if len(jsonData) == 0 {
		return []entity.NoteAttachment{}
	}

	var attachments []entity.NoteAttachment
	if err := json.Unmarshal(jsonData, &attachments); err != nil {
		return []entity.NoteAttachment{}
	}

	return attachments
}

File: crm/src/main/java/serp/project/crm/infrastructure/store/mapper/NoteMapper.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.crm.infrastructure.store.mapper;

import org.springframework.stereotype.Component;
import serp.project.crm.core.domain.entity.NoteEntity;
import serp.project.crm.infrastructure.store.model.NoteModel;
import lombok.RequiredArgsConstructor;
import java.util.List;

@Component
@RequiredArgsConstructor
public class NoteMapper extends BaseMapper {

    public NoteEntity toEntity(NoteModel model) {
        if (model == null) {
            return null;
          }
        return NoteEntity.builder()
                .id(model.getId())
                .tenantId(model.getTenantId())
                .entityType(model.getEntityType())
                .entityId(model.getEntityId())
                .content(model.getContent())
                .createdAt(toTimestamp(model.getCreatedAt()))
                .updatedAt(toTimestamp(model.getUpdatedAt()))
                .createdBy(model.getCreatedBy())
                .updatedBy(model.getUpdatedBy())
                .build();
    }

    public NoteModel toModel(NoteEntity entity) {
        if (entity == null) {
            return null;
        }
        return NoteModel.builder()
                .id(entity.getId())
                .tenantId(entity.getTenantId())
                .entityType(entity.getEntityType())
                .entityId(entity.getEntityId())
                .content(entity.getContent())
                .createdAt(toLocalDateTime(entity.getCreatedAt()))
                .updatedAt(toLocalDateTime(entity.getUpdatedAt()))
                .createdBy(entity.getCreatedBy())
                .updatedBy(entity.getUpdatedBy())
                .build();
    }

    public List<NoteEntity> toEntityList(List<NoteModel> models) {
        if (models == null) {
            return null;
        }
        return models.stream().map(this::toEntity).toList();
    }
}
